deps/undici/src/lib/observable.js

```diff
@@ -0,0 +1,802 @@
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+class Observable {
+  static from(value) {
+    if (value instanceof Observable) {
+      return value;
+    }
+    if (value && typeof value.then === 'function') {
+      return new Observable((subscriber) => {
+        value.then(
+          (value) => {
+            subscriber.next(value);
+            subscriber.complete();
+          },
+          (error) => {
+            subscriber.error(error);
+          }
+        );
+      });
+    }
+    if (Symbol.asyncIterator in value) {
+      return new Observable(async (subscriber) => {
+        const iterator = value[Symbol.asyncIterator]();
+        try {
+          while (subscriber.isActive) {
+            const { value, done } = await iterator.next();
+            if (done) {
+              subscriber.complete();
+              return;
+            }
+            subscriber.next(value);
+          }
+        } catch (error) {
+          subscriber.error(error);
+        } finally {
+          iterator.return?.();
+        }
+      });
+    }
+    if (Symbol.iterator in value) {
+      return new Observable((subscriber) => {
+        const iterator = value[Symbol.iterator]();
+        try {
+          while (subscriber.isActive) {
+            const { value, done } = iterator.next();
+            if (done) {
+              subscriber.complete();
+              return;
+            }
+            subscriber.next(value);
+          }
+        } catch (error) {
+          subscriber.error(error);
+        } finally {
+          iterator.return?.();
+        }
+      });
+    }
+    throw new TypeError('Value is not observable');
+  }
+
+  #start = null;
+
+  constructor(start) {
+    this.#start = start;
+  }
+
+  subscribe(fnOrObserver, options) {
+    if (this.#start) {
+      const partialObserver =
+        typeof fnOrObserver === 'function'
+          ? { next: fnOrObserver }
+          : fnOrObserver;
+      const observer = {
+        next: noop,
+        error: reportError,
+        complete: noop,
+        ...(partialObserver ?? {}),
+      };
+      const subscriber = new Subscriber(options?.signal, observer);
+      this.#start(subscriber);
+    }
+  }
+
+  forEach(callback, options) {
+    return new Promise((resolve, reject) => {
+      const ac = new AbortController();
+      const signal = options?.signal
+        ? AbortSignal.all([ac.signal, options.signal])
+        : ac.signal;
```

Comment:
Should be more performant to de-reference signals once:

```suggestion
      const ac = new AbortController();
      const acSignal = ac.signal;
      const optionsSignal = options?.signal;
      const signal = optionsSignal
        ? AbortSignal.all([acSignal, optionsSignal])
        : acSignal;
```